### ISSUE TITLE
Failure to run embedded JDK's java binary to launch Bazel on Windows

### ISSUE THREAD

**Issue by cj741** [bazelbuild/bazel#8875]
> ATTENTION! Please read and follow:
> - if this is a _question_ about how to build / test / query / deploy using Bazel, ask it on StackOverflow instead: https://stackoverflow.com/questions/tagged/bazel
> - if this is a _discussion starter_, send it to bazel-discuss@googlegroups.com
> - if this is a _bug_ or _feature request_, fill the form below as best as you can.

### Description of the problem / feature request:

After I installed bazel 0.27.1, I try to run the command "bazel", it show the error
"FATAL: ExecuteProgram(C:\Users\Administrator/_bazel_Administrator/install/9470fb600225157f1e34c45c6d4dc834/_embedded_binaries/embedded_tools/jdk/bin/java.exe) failed: 6"
but when move into the path "C:\Users\Administrator/_bazel_Administrator/install/9470fb600225157f1e34c45c6d4dc834/_embedded_binaries/embedded_tools/jdk/bin/", it is ok to run "bazel", do you have idea on it? and how to solve it.
Thanks~

**Comment by aiuto** [comment]
Does this still happen with the more recent Bazel release - 0.28,0?

**Comment by laszlocsomor** [comment]
@cj741 : What Windows version do you use? Did you install Bazel from Chocolatey? Can you try running it from a non-Administrator command prompt -- does that change anything? Is your machine 64 bit or 32 bit?

**Comment by cj741** [comment]
@laszlocsomor Thanks for your response. It works well with the non-Adminstrator command prompt! whats wrong with that? I want to build tensorflow, I think it need the Administrator command promp, do you have idea how to works on Administrator?

**Comment by laszlocsomor** [comment]
I'm glad it works! It's a mystery why Admin shell makes a difference -- let's try to find out. What Windows version do you use? Is it 64 bit or 32 bit?

As best I know, you don't need Admin shell to build TensorFlow. In fact you should never need Admin shell when using Bazel. Can you try building TF in a non-Admin shell now?

**Comment by yuanbaopapa** [comment]
I have the same issue as cj741, but changing to non-Admin console did not help. my bazel version is 0.28.1 ,my OS is windows 7 64 bit. I installed bazel from github binary.Here is my error msg:
FATAL: ExecuteProgram(C:\Users\host/_bazel_host/install/a4556c0a24a7f5c72797f234cbf35797/_embedded_binaries/embedded_tools/jdk/bin/java.exe) failed:

If I enter the directory "C:\Users\host/_bazel_host/install/a4556c0a24a7f5c72797f234cbf35797/_embedded_binaries/embedded_tools" and then run bazel.exe there ,no error occured.However this workaround did not help to build tensorflow

**Comment by yuanbaopapa** [comment]
bazel version 0.24.1 works, 0.26.1 and 0.28.1 failed with error msg as above

**Comment by yuanbaopapa** [comment]
I start bazel in debugger as bazel --batch ,and set a break at CreateProcessW where bazel will create java process before emitting the error msg,and get the lCommand parameter for bazel as following, guess there may be something wrong with the lCommand parameter :

"C:\Users\host\_bazel_host\install\a4556c0a24a7f5c72797f234cbf35797\_embedded_binaries\embedded_tools\jdk\bin\java.exe"  -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=c:\\\users\\\host\\\_bazel_host\\\u5mzpase -XX:+UseParallelOldGC --add-opens=java.base/java.nio=ALL-UNNAMED --add-opens=java.base/java.lang=ALL-UNNAMED -Xverify:none -Djava.util.logging.config.file=c:\\\users\\\host\\\_bazel_host\\\u5mzpase\\\javalog.properties -Dcom.google.devtools.build.lib.util.LogHandlerQuerier.class=com.google.devtools.build.lib.util.SimpleLogHandler$HandlerQuerier -XX:-MaxFDLimit -Djava.library.path=C:\\\Users\\\host\\\_bazel_host\\\install\\\a4556c0a24a7f5c72797f234cbf35797\\\_embedded_binaries\\\embedded_tools\\\jdk\\\bin;C:\\\Users\\\host\\\_bazel_host\\\install\\\a4556c0a24a7f5c72797f234cbf35797\\\_embedded_binaries\\\embedded_tools\\\jdk\\\bin\\\server;C:\\\Users\\\host\\\_bazel_host\\\install\\\a4556c0a24a7f5c72797f234cbf35797\\\_embedded_binaries -Dfile.encoding=ISO-8859-1 -Dbazel.windows_unix_root=C:\msys64\usr\bin\bash.exe -jar C:\\\Users\\\host\\\_bazel_host\\\install\\\a4556c0a24a7f5c72797f234cbf35797\\\_embedded_binaries\\\A-server.jar --batch --connect_timeout_secs=30 --output_user_root=c:\users\host\_bazel_host --install_base=c:\users\host\_bazel_host\install\a4556c0a24a7f5c72797f234cbf35797 --install_md5=a4556c0a24a7f5c72797f234cbf35797 --output_base=c:\users\host\_bazel_host\u5mzpase --workspace_directory= "--default_system_javabase=C:/Program Files/Java/jdk-11.0.1" --deep_execroot --expand_configs_in_place --idle_server_tasks --noexperimental_oom_more_eagerly --experimental_oom_more_eagerly_threshold=100 --write_command_log --nowatchfs --nofatal_event_bus_exceptions --client_debug=false --host_jvm_args=-Dbazel.windows_unix_root=C:\msys64\usr\bin\bash.exe --product_name=Bazel --option_sources=batch:

**Comment by yuanbaopapa** [comment]
after comparing lpCommandLine of CreateProcessW in bazel version 0.28.1 with lpCommandLine of CreateProcessA in bazel version 0.24.1, the only difference between them is that version 0.24.1 has another parameter to invoke embeded java.exe: --noincompatible_windows_style_arg_escaping

**Comment by yuanbaopapa** [comment]
So there may be some bug about dealing with windows arg escaping for versions above 0.24.1

**Comment by laszlocsomor** [comment]
@yuanbaopapa : Thanks for debugging! Nothing looks wrong with the command line. The `--[no]incompatible_windows_style_arg_escaping` flag might be the culprit, not sure. I'll try to repro on Windows Server 2008 (equivalent to Win 7).

**Comment by laszlocsomor** [comment]
Bazel 0.28.1 works for me in a blank workspace (just empty `WORKSPACE` and `BUILD` files) on Windows Server 2008 R2.

@yuanbaopapa : Could you try this in cmd.exe please:
```
md foo

cd foo

type nul > WORKSPACE

type nul > BUILD

c:\path\to\bazel-0.28.1-windows-x86_64.exe build //...
```

Does that work?

**Comment by laszlocsomor** [comment]
Ah, I managed to repro *some* breakage:

```
C:\Users\laszlocsomor\ws>..\bazel-0.28.1-windows-x86_64.exe --batch build //...
(...)
FATAL: ExecuteProgram(C:\Users\laszlocsomor/_bazel_laszlocsomor/install/a4556c0a24a7f5c72797f234cbf35797/_embedded_binaries/embedded_tools/jdk/bin/java.exe) failed: ERROR: src/main/native/windows/process.cc(184): CreateProcessW("C:\Users\laszlocsomor\_bazel_laszlocsomor\install\a4556c0a24a7f5c72797f234cbf35797\_embedded_binaries\embedded_tools\jdk\bin\java.exe"  -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=c:\\\user(...)): Insufficient system resources exist to complete the requested service.
```

So the `--batch` mode does make a difference, but my error message is different: "Insufficient system resources exist to complete the requested service.". This error looks like https://github.com/bazelbuild/bazel/pull/8793, whose fix is NOT in Bazel 0.28.1, because 0.28's baseline is https://github.com/bazelbuild/bazel/commit/2e374a9c6e3d4ed71f0145de287c4b2fe43c76d6 and the bugfix was merged later as https://github.com/bazelbuild/bazel/commit/c4a469ad4cb689962e0b15c71b770cc87753300f.

If I try Bazel from HEAD (e.g. the [last nightly](https://buildkite.com/organizations/bazel/pipelines/bazel-at-head-plus-downstream/builds/1124/jobs/5f303771-d340-412b-a96b-02727991dcc7/artifacts/e57132ee-bbe8-4c93-8451-a55c89d7d571)), **it works**.

@yuanbaopapa : Could you please try:
1. download https://buildkite.com/organizations/bazel/pipelines/bazel-at-head-plus-downstream/builds/1124/jobs/5f303771-d340-412b-a96b-02727991dcc7/artifacts/e57132ee-bbe8-4c93-8451-a55c89d7d571
2. rename the file to `bazel-nightly.exe`
3. use that instead of release Bazel to build what was failing

Does that work?

**Comment by jin** [comment]
This seems to have been fixed in @laszlocsomor's latest comment. Closing for now, please ping to reopen this issue if you're still seeing the problem.
